### PR TITLE
Incorporation of state level electricity grid mix for performing LCA calculations

### PR DESCRIPTION
**Pull request by TJTapajyoti**
- This requires a new electricity grid mix file with state level information that has been added to the data branches - tiny-data, medium-data, national-data. Is it required to be added to different repo branches?
- config file requires the name of the proper electricity grid mix file. There are two - national level and state level. The current code in issue 83 will not work if we use the national level grid mix. 
- @rjhanes Do we need to different versions of the code? If not, we can remove the old data inventory for national level grid mix from the repo. For now I have kept both. 
- The name of the file in the config file used for running the code should be changed. - https://nrel-my.sharepoint.com/personal/aeberle_nrel_gov/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Faeberle%5Fnrel%5Fgov%2FDocuments%2FCELAVI%20LDRD%20Team%20Folder%2Fdata%2Fconfigs&FolderCTID=0x012000B395385DB8CF0449BF510272F1FACBCC

### CODE REVIEW

File: celavi/pylca_celavi/pylca_opt_foreground.py
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_foreground.py:19]: I think we'll probably want to compare results using the national grid to the state grid at some point. Can you add some functionality to use either grid mix, either through a Boolean flag or by defaulting to the national grid if the grid mix file doesn't have a state column?

File: celavi/pylca_celavi/des_interface.py
Comment (rjhanes) [inline celavi/pylca_celavi/des_interface.py:227]: Whenever an empty results data frame signifies an error with the optimization, add in a warning that we can use for debugging. There's a few statements like this throughout the code.

**Review by rjhanes — CHANGES_REQUESTED** [review]
I had to make some changes to this branch so it would run on my machine. See branch `issue-83-tiny` on the data repo for the config files and input files I was using.

I'm not seeing environmental impacts for any stages other than manufacturing and transportation when I run the code, and I'm getting repeated messages saying there is zero final demand for processes that should have final demand (coarse grinding onsite, rotor teardown, landfilling). This bug has occurred before so hopefully the fix will be simple.

I'd like additional functionality so we can switch between the national and state grid mixes and compare results between the two. See my comment in the code for suggestions.

**Comment by TJTapajyoti** [comment]
Due to an error in the material name, the tiny and medium data branch had an old version of the inventory file. I have updated the tiny data, medium data with the new version of the file such that all stages LCIA are performed.

File: celavi/pylca_celavi/des_interface.py
Comment (TJTapajyoti) [reply celavi/pylca_celavi/des_interface.py:225]: Added statements for optimization failure warning.

File: celavi/pylca_celavi/pylca_opt_foreground.py
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_foreground.py:19]: I have started implementing this functionality - But there are some issues due to many parts of the code requiring change......Working through them.

Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_foreground.py]: Incorporated

File: celavi/pylca_celavi/des_interface.py
Comment (rjhanes) [inline celavi/pylca_celavi/des_interface.py:261]: I updated this print statement so it's more informative and can be used for debugging. However now, with either the national or state grid mix, I'm getting output messages like:

```
Final demand for 2033 next use glass fiber is zero
LCIA optimization failed for glass fiber at next use in 2033
```

So I don't think the if/else statement structures here are correct, because it's saying the optimization failed when no optimization should have run anyway.

File: celavi/pylca_celavi/insitu_emission.py
Comment (rjhanes) [inline celavi/pylca_celavi/insitu_emission.py:207]: I updated the other optimization failed statements as well, for easier debugging

**Comment by rjhanes** [comment]
I did get a couple failed optimizations when running without the shortcut calculations. With the state-level grid mix:

```
Pylca-opt-background emission optimization failed for glass fiber at landfilling in 2033
```

With the national grid mix, there's a consistent problem with the manufacturing:

```
Pylca-opt-background emission optimization failed for glass fiber at manufacturing in 2017
Pylca-opt-background emission optimization failed for glass fiber at manufacturing in 2017
```

This appears to be happening whenever there is manufacturing

**Review by rjhanes — APPROVED** [review]
The "LCIA optimization failed" print statement is being shown when final demand for a material in a stage is zero, meaning the calculations shouldn't be run anyway. That statement should either be moved or maybe deleted - there are many other warnings for failed optimizations and it's not clear if this one is necessary.

When I'm running without shortcut calculations, I am getting a few other failed optimizations - see my other comment. Landfilling fails sporadically when using the state level grid mix and manufacturing fails consistently when using the national grid mix. Since this is the first time I'm running with the failure warnings, I don't know if these are to be expected or not. I recommend testing and looking into the LCI results to see if there's actually a problem.

I'm approving, so you can merge in once you've moved or deleted the LCIA optimization failed print statement referenced above. If there is any additional coding needed to deal with the landfilling and manufacturing LCI failures, the I would like to re-review before merging in.

File: celavi/pylca_celavi/insitu_emission.py
Comment (TJTapajyoti) [reply celavi/pylca_celavi/insitu_emission.py:207]: I updated many print statements in this fashion for easier debugging.

File: celavi/pylca_celavi/des_interface.py
Comment (TJTapajyoti) [reply celavi/pylca_celavi/des_interface.py]: I actually made some changes to the code to have print statements for explicit cases of 

1. No listing in the LCA inventory, thus returning empty dataframe. 
2. No final demand, flows are zero. 
3. Actual failure of optimization due to some error. 
The case for next use and in use of materials falls in category 1. Thus for that category, we get two print statements - 
a. There is not listing of LCA inventory for this material and stage and b, the LCIA code returned an empy dataframe. 

These changes have been incorporated in PR for issue #113 

> The "LCIA optimization failed" print statement (line 261 in des_interface) is being shown when final demand for a material in a stage is zero, meaning the calculations shouldn't be run anyway. That statement should either be moved or maybe deleted - there are many other warnings for failed optimizations and it's not clear if this one is necessary.
> 
> When I'm running without shortcut calculations, I am getting a few other failed optimizations - see my other comment. Landfilling fails sporadically when using the state level grid mix and manufacturing fails consistently when using the national grid mix. Since this is the first time I'm running with the failure warnings, I don't know if these are to be expected or not. I recommend testing and looking into the LCI results to see if there's actually a problem.
> 
> I'm approving, so you can merge in once you've moved or deleted the LCIA optimization failed print statement referenced above. If there is any additional coding needed to deal with the landfilling and manufacturing LCI failures, the I would like to re-review before merging in.

1. When the final demand is zero, I have added and explicit statement that just says final demand is zero and nothing happens. For cases like in use, next use, final demand is not zero. There is a final demand but due to no listing in the LCA Inventory, no calculations occur. I have added an explicit print statement highlighting this situation. 
2.

**Comment by TJTapajyoti** [comment]
- I performed two full runs with tiny data branch and issue 83 branch of the code repo and did not face any optimization errors. 
- Print statements were corrected and easily understandable now. 
- Functionality of the Reeds importer class and manipulation of Reeds manually edited data have been merged with changes to this branch, The error for the optimization may have been rectified due to this addition. 
- @rjhanes @akey7 I will request a complete test run on your machines before merging this with develop.

**Review by akey7 — APPROVED** [review]
Looks good to me.

**Pull request merged**
